# Customer Segmentation & Churn Pattern Analytics in European Banking

## Notebook 01: Data Ingestion & Validation

---

### Project Overview

The first step in any data analytics project is to ensure that the dataset is accurate, consistent, and suitable for analysis. This notebook focuses on loading the banking customer dataset and validating key business-critical fields before moving on to data cleaning and exploratory analysis.

Performing these validation checks helps identify potential data quality issues early and establishes confidence in the reliability of the dataset.

---

## Objectives

The objectives of this notebook are to:

A. **Load the dataset** into a Pandas DataFrame for analysis.

B. **Validate engagement and product fields** to ensure the data accurately represents customer interactions with the bank.

C. **Ensure binary variables consistency** by verifying that all binary features contain only valid values.

D. **Confirm churn labeling accuracy** by validating the target variable used throughout the project.

---

## Expected Outcome

By the end of this notebook, the dataset will be successfully loaded and validated, ensuring it is reliable and ready for the **Data Cleaning & Preparation** phase.

---

In [1]:
# Import Required Libraries
import pandas as pd
import numpy as np

In [2]:
# Load the Dataset
file_path = "../data/European_Bank.csv"
df = pd.read_csv(file_path)

# Preview the first five rows
df.head()

,Year,CustomerId,Surname,CreditScore,Geography,Gender,Age,Tenure,Balance,NumOfProducts,HasCrCard,IsActiveMember,EstimatedSalary,Exited
0,2025,15634602,Hargrave,619,France,Female,42,2,0.00,1,1,1,101348.88,1
1,2025,15647311,Hill,608,Spain,Female,41,1,83807.86,1,0,1,112542.58,0
2,2025,15619304,Onio,502,France,Female,42,8,159660.80,3,1,0,113931.57,1
3,2025,15701354,Boni,699,France,Female,39,1,0.00,2,0,0,93826.63,0
4,2025,15737888,Mitchell,850,Spain,Female,43,2,125510.82,1,1,1,79084.10,0


---
# A. Initial Dataset Inspection

In [3]:
print(f"Rows    : {df.shape[0]}")
print(f"Columns : {df.shape[1]}")

# Display column names
print(f"column names  : {df.columns.tolist()}")

Rows    : 10000
Columns : 14
column names  : ['Year', 'CustomerId', 'Surname', 'CreditScore', 'Geography', 'Gender', 'Age', 'Tenure', 'Balance', 'NumOfProducts', 'HasCrCard', 'IsActiveMember', 'EstimatedSalary', 'Exited']


In [4]:
# data information
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Year             10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  str    
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  str    
 5   Gender           10000 non-null  str    
 6   Age              10000 non-null  int64  
 7   Tenure           10000 non-null  int64  
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(2), int64(9), str(3)
memory usage: 1.1 MB


In [5]:
# Statistical Summary 
df.describe().T

,count,mean,std,min,25%,50%,75%,max
Year,10000.0,2.025000e+03,0.000000,2025.00,2025.00,2.025000e+03,2.025000e+03,2025.00
CustomerId,10000.0,1.569094e+07,71936.186123,15565701.00,15628528.25,1.569074e+07,1.575323e+07,15815690.00
CreditScore,10000.0,6.505288e+02,96.653299,350.00,584.00,6.520000e+02,7.180000e+02,850.00
Age,10000.0,3.892180e+01,10.487806,18.00,32.00,3.700000e+01,4.400000e+01,92.00
Tenure,10000.0,5.012800e+00,2.892174,0.00,3.00,5.000000e+00,7.000000e+00,10.00
Balance,10000.0,7.648589e+04,62397.405202,0.00,0.00,9.719854e+04,1.276442e+05,250898.09
NumOfProducts,10000.0,1.530200e+00,0.581654,1.00,1.00,1.000000e+00,2.000000e+00,4.00
HasCrCard,10000.0,7.055000e-01,0.455840,0.00,0.00,1.000000e+00,1.000000e+00,1.00
IsActiveMember,10000.0,5.151000e-01,0.499797,0.00,0.00,1.000000e+00,1.000000e+00,1.00
EstimatedSalary,10000.0,1.000902e+05,57510.492818,11.58,51002.11,1.001939e+05,1.493882e+05,199992.48


---
## Looking for Missing Values & Duplicate Records

Before diving into the analysis, it's important to make sure the dataset is complete and free from duplicate records.

- Missing values can affect calculations, visualizations, and model performance, while duplicate records may introduce bias by counting the same customer more than once.
- In this step, we'll verify that the dataset is clean and identify any issues that need to be addressed during the data preparation phase.

In [6]:
# Count missing values in each column
missing_values = df.isnull().sum()

# Create a summary table
missing_summary = pd.DataFrame({
    "Missing Values": missing_values,
    "Missing Percentage (%)": (missing_values / len(df) * 100).round(2)
})

# Display only columns with missing values
missing_summary[missing_summary["Missing Values"] > 0]

,Missing Values,Missing Percentage (%)


### Observation

The summary table above highlights any columns containing missing values.

As no rows are returned, the dataset is complete and no immediate action is required before moving to the next stage of analysis.

In [7]:
# Duplicate Records

duplicate_count = df.duplicated().sum()

print(f"Total Duplicate Records : {duplicate_count}")

Total Duplicate Records : 0


### Observation

A duplicate count of **0** indicates that every customer record is unique.

If duplicate records are found, they will be investigated and removed during the data cleaning phase.

---

# B. Validating the Dataset Structure

Now that we've explored the dataset, it's time to verify whether its structure aligns with our expectations.

We'll check:

- The total number of rows and columns
- Whether the expected columns are present
- If any columns are missing or unexpected

Performing these checks early helps catch structural issues before they affect downstream analysis.

In [8]:
# Dataset Structure Validation

expected_columns = ["Year","CustomerId","Surname","CreditScore","Geography","Gender","Age","Tenure",
                    "Balance","NumOfProducts","HasCrCard","IsActiveMember","EstimatedSalary","Exited"
]

actual_columns = df.columns.tolist()

missing_columns = set(expected_columns) - set(actual_columns)
extra_columns = set(actual_columns) - set(expected_columns)

print("DATASET STRUCTURE VALIDATION")

print(f"Rows                : {df.shape[0]}")
print(f"Columns             : {df.shape[1]}")
print(f"Expected Columns    : {len(expected_columns)}")
print(f"Available Columns   : {len(actual_columns)}")

print("\nSchema Validation")

print(f"Missing Columns     : {missing_columns if missing_columns else 'None'}")
print(f"Unexpected Columns  : {extra_columns if extra_columns else 'None'}")

if not missing_columns and not extra_columns:
    print("\n Dataset schema validation passed.")
else:
    print("\n Dataset schema validation failed.")

DATASET STRUCTURE VALIDATION
Rows                : 10000
Columns             : 14
Expected Columns    : 14
Available Columns   : 14

Schema Validation
Missing Columns     : None
Unexpected Columns  : None

 Dataset schema validation passed.


---
# Validating the Year Column

The dataset contains a **Year** column, which appears to represent the year in which the customer data was recorded.

Before proceeding, we'll verify whether this column contains a single consistent value or multiple years. This helps determine whether the feature is useful for analysis or can be removed during the data cleaning phase.

In [9]:
# Year Column Validation
print(df["Year"].value_counts().sort_index())

Year
2025    10000
Name: count, dtype: int64


### Observation

The **Year** column contains a single unique value (**2025**) across all 10,000 customer records.

Since there is no variation in this feature, it does not contribute meaningful information for customer segmentation or churn analysis. The column will be removed during the data cleaning phase while preserving the original raw dataset.

---

# Validating Business-Critical Fields

The reliability of business insights depends on the quality and consistency of the underlying data.

In this section, we validate key business-related variables that play a central role in customer churn analysis.

The validation focuses on:

- Credit card ownership
- Customer activity status
- Number of bank products owned
- Churn labels

These checks help ensure that the data accurately represents real-world customer information before moving on to exploratory analysis.

In [10]:
# Validate Credit Card Ownership
credit_card_distribution = (
    df["HasCrCard"]
      .value_counts()
      .sort_index()
      .rename_axis("Has Credit Card")
      .reset_index(name="Customer Count")
)

credit_card_distribution

,Has Credit Card,Customer Count
0,0,2945
1,1,7055


In [11]:
# Percentage Distribution

credit_card_distribution["Percentage (%)"] = (
    credit_card_distribution["Customer Count"] / len(df) * 100
).round(2)

credit_card_distribution

,Has Credit Card,Customer Count,Percentage (%)
0,0,2945,29.45
1,1,7055,70.55


### Observation

The **HasCrCard** feature contains only the expected binary values (**0** and **1**), confirming that the data is logically consistent.

Among the **10,000 customers**, **7,055 (70.55%)** own a bank-issued credit card, while **2,945 (29.45%)** do not. This indicates that credit card ownership is common within the customer base and may serve as a useful variable when analyzing customer behavior and churn patterns in later stages of the project.

---

## Customer Activity Status

The **IsActiveMember** column indicates whether a customer is actively using the bank's services.

Expected values:

- **0** → Inactive Customer
- **1** → Active Customer

Since customer engagement is closely related to retention, validating this feature is essential before investigating churn patterns.

In [12]:
# Validate Customer Activity Status

activity_distribution = ( 
    df["IsActiveMember"].value_counts().sort_index().rename_axis("Active Member").reset_index(name="Customer Count")
)

activity_distribution["Percentage (%)"] = (
    activity_distribution["Customer Count"] / len(df) * 100).round(2)

activity_distribution

,Active Member,Customer Count,Percentage (%)
0,0,4849,48.49
1,1,5151,51.51


In [13]:
# Validate Binary Consistency

valid_values = {0, 1}

unique_values = set(df["IsActiveMember"].unique())

if unique_values.issubset(valid_values):
    print("IsActiveMember contains only valid binary values.")
else:
    print("Invalid values found:", unique_values)

IsActiveMember contains only valid binary values.


### Observation

The **IsActiveMember** feature passed the binary consistency check, confirming that it contains only the expected values (**0** and **1**).

Out of **10,000 customers**, **5,151 (51.51%)** are active members, while **4,849 (48.49%)** are inactive. The distribution is relatively balanced, providing a strong foundation for investigating whether customer engagement influences churn behavior in later stages of the analysis.

---

## Product Ownership Validation

The **NumOfProducts** column represents the number of banking products currently used by each customer.

Examples of banking products include savings accounts, current accounts, credit cards, loans, and investment services.

Before using this feature for customer segmentation and churn analysis, it is important to verify that the values are valid and understand how customers are distributed across different product categories.

In [14]:
# Product Ownership Distribution

product_distribution = (
    df["NumOfProducts"]
    .value_counts()
    .sort_index()
    .reset_index(name="Customer Count")
    .rename(columns={"index": "Number of Products"})
)

product_distribution["Percentage (%)"] = (
    product_distribution["Customer Count"] / len(df) * 100
).round(2)

product_distribution

,NumOfProducts,Customer Count,Percentage (%)
0,1,5084,50.84
1,2,4590,45.90
2,3,266,2.66
3,4,60,0.60


In [15]:
# Unique Product Counts
list(map(int, sorted(df["NumOfProducts"].unique())))

[1, 2, 3, 4]

### Observation

The **NumOfProducts** feature contains four valid product categories (**1**, **2**, **3**, and **4**), with no unexpected or invalid values detected.

Most customers own either **one** or **two** banking products, accounting for **96.74%** of the customer base. Only a small proportion of customers use **three (2.66%)** or **four (0.60%)** products.

This distribution appears realistic and confirms that the feature is suitable for further analysis, particularly when examining the relationship between product ownership and customer churn.

---

## C. Binary Variables Consistency Check

Binary variables represent two possible outcomes and are expected to contain only the values **0** and **1**.

To ensure data integrity, each binary field is validated for consistency before proceeding with further analysis.

The binary variables included in this validation are:

- **HasCrCard**
- **IsActiveMember**
- **Exited**

In [16]:
# Binary Variables Consistency Check

binary_columns = [
    "HasCrCard",
    "IsActiveMember",
    "Exited"
]

validation_results = []

for column in binary_columns:

    unique_values = sorted(df[column].unique())

    validation_results.append({
        "Column": column,
        "Unique Values": unique_values,
        "Validation Status": "Passed" if set(unique_values) == {0, 1} else "Failed"
    })

validation_summary = pd.DataFrame(validation_results)

validation_summary

,Column,Unique Values,Validation Status
0,HasCrCard,"[0, 1]",Passed
1,IsActiveMember,"[0, 1]",Passed
2,Exited,"[0, 1]",Passed


---

## D. Churn Label Validation

The **Exited** column is the target variable for this project and indicates whether a customer has left the bank.

The expected values are:

- **0** → Customer Retained
- **1** → Customer Churned

Before using this variable for further analysis, it is important to verify that the labels are correctly encoded and examine the overall distribution of retained and churned customers.

In [17]:
# Churn Distribution

churn_distribution = (
    df["Exited"]
    .value_counts()
    .sort_index()
    .reset_index(name="Customer Count")
    .rename(columns={"index": "Exited"})
)

churn_distribution["Customer Status"] = [
    "Retained",
    "Churned"
]

churn_distribution["Percentage (%)"] = (
    churn_distribution["Customer Count"] / len(df) * 100
).round(2)

churn_distribution

,Exited,Customer Count,Customer Status,Percentage (%)
0,0,7963,Retained,79.63
1,1,2037,Churned,20.37


In [18]:
# Overall Churn Rate

churn_rate = (
    df["Exited"].mean() * 100
).round(2)

print(f"Overall Churn Rate: {churn_rate}%")

Overall Churn Rate: 20.37%


In [19]:
# Validate Churn Labels

sorted(df["Exited"].unique())

[np.int64(0), np.int64(1)]

### Observation

The **Exited** column was successfully validated and contains only the expected binary values (**0** for retained customers and **1** for churned customers), confirming that the target variable is correctly encoded.

Out of **10,000 customers**, **7,963 (79.63%)** were retained, while **2,037 (20.37%)** exited the bank. This indicates that approximately one in every five customers has churned.

The overall churn rate of **20.37%** serves as the baseline performance metric for this project and will be used as a reference when evaluating churn patterns across different customer segments in the subsequent analysis.

---

# Conclusion

The data ingestion and validation process was completed successfully, confirming that the dataset is reliable for further analysis.

### Key Findings

- The dataset was imported successfully and is ready for analysis.
- Customer engagement fields (`HasCrCard` and `IsActiveMember`) contain valid and consistent values.
- Product ownership (`NumOfProducts`) includes only expected product categories.
- All binary variables (`HasCrCard`, `IsActiveMember`, and `Exited`) passed the consistency checks.
- The churn labels are correctly encoded, with an overall churn rate of **20.37%**.

With the initial validation complete, the dataset is ready for the next phase: **Data Cleaning & Preparation**, where unnecessary features will be removed, data types reviewed, and new analytical features will be created to support deeper customer segmentation and churn analysis.